### ISSUE TITLE
redesign copy and paste workflow

### ISSUE THREAD

**Issue by parafin** [darktable-org/darktable#6586]
Split from #6564:

I find current copy-paste system unintuitive. I think there should be 3 buttons instead of current 4:

1. copy (without dialog)
2. filter (dialog where you select the iops to transfer)
3. paste (without dialog)

The current situation where iop selection is bundled with copy and paste operations results in complexity and non-transparent behaviour without adding anything.

Now, filter button in the proposed scheme is where all the magic happens. It should display its status (all selected, only safe, custom selection), and we can have presets for it. By default it should be only safe modules (whatever the list might be), and reset action should reset it to that state. Not sure if its state should be saved across dt restarts, probably not, presets should be used instead. I think it will fix all use-cases without the need for yet another preference option (really, you gotta stop adding them).

Answering to @TurboGit comment:
> > I think there should be 3 buttons instead of current 4:
> 
> Not sure, one of my workflow is:
> 
> * copy all
> * select some pictures
> * paste parts
> * select some other pictures
> * paste parts (not the same iop than above)

It will change to:

* copy
* select some pictures
* filter
* paste
* select some other pictures
* filter
* paste

You can also paste the same iops several times (to different selections) without going through the selection dialog.

> 
> The other is:
> 
> * copy parts
> * select some pictures
> * paste all

It will change to:

* copy
* select some pictures
* filter
* paste

Notice how it's not different at all from the first use-case, meaning it's now simpler to understand. And this use-case is exactly the problem I'm having with current design - what does "paste all" pastes here? Only iops selected in "copy parts" action? But then if I do "paste parts" I can then click "select all" in appeared dialog and paste more iops than "paste all" does. What does "all" mean then? At the very least it's a poor choice of words.

**Comment by elstoc** [comment]
I still think the copy and select/filter should be combined - it makes it easier when using it with shortcuts outside of the lighttable. You would have two forms of copy (with or without the dialog) and then only a single form of paste that copies whatever was selected. You can customise the copy-without-dialog using a default set of "all" or "safe" within the lighttable module. Similar to how you select overwrite or append modes.

**Comment by parafin** [comment]
Well, we can have shortcut that calls both copy and filter, why not. No need to change anything in the design to do that IMHO. And yes, selected filter should be persistent across view changes.

**Comment by elstoc** [comment]
I think that's the more common use case so it should be simple. Is it really a common use case to paste different subsets of modules from a single source to different destinations (which feels like the only time you'd need to re-filter)?

**Comment by parafin** [comment]
What is the problem? That you will need to press one more button compared to current situation? You can and should use shortcuts if you're concerned with optimizing your workflow, and we already covered that it can be a single shortcut even if there are 2 buttons in the GUI.

Ability to re-filter is probably not very important, I agree, but ability to re-use the filter is IMHO quite powerful - e.g. you can copy just white balance several times (from different sources) without having to go through the dialog several times.

**Comment by elstoc** [comment]
> you can copy just white balance several times (from different sources) without having to go through the dialog several times.

Oh I hadn't realised that's what you meant by your filter dialog - choosing a list of modules rather than a list of instances from a specific image.

**Comment by parafin** [comment]
> > you can copy just white balance several times (from different sources) without having to go through the dialog several times.
> 
> Oh I hadn't realised that's what you meant by your filter dialog - choosing a list of modules rather than a list of instances from a specific image.

Dialog should probably have 2 modes - display just iops from currently copied image, or list of all iops. And selection can be saved as preset for history lib module.

**Comment by dterrahe** [comment]
my 2c: as an excel/libreoffice calc user, what makes sense to me is
copy
paste special...
where second step combines selecting subset of properties and pasting.
I'm not sure what splitting this into two steps would add. The property selection dialog would retain the previous subset, so you can do multiple pastes of the same modules quickly by just pressing enter when the dialog comes up.

**Comment by parafin** [comment]
Multi-instance though creates a problem here... Not sure how to resolve it.

**Comment by parafin** [comment]
> my 2c: as an excel/libreoffice calc user, what makes sense to me is
> copy
> paste special...
> where second step combines selecting subset of properties and pasting.
> I'm not sure what splitting this into two steps would add. The property selection dialog would retain the previous subset, so you can do multiple pastes of the same modules quickly by just pressing enter when the dialog comes up.

Well, that's a question if this scenario is OK. I suspect we have 4 buttons right now, because someone wanted a possibility not to re-display the selection dialog, and that's what I tried to optimize for too. Your way we can remove "copy parts" button and call it a day, which would be a step in the right direction IMHO.

Splitting copy and paste into 3 steps instead of 2 creates a more powerful instrument while keeping things simple and transparent in how they work. Though there are problems with this approach, namely: one more GUI button press in some use-cases and multi-instance not being covered well.

**Comment by dterrahe** [comment]
I don't mind any outcome, so don't take this as advocating, but the two button "paste special" approach does allow to make adjustments quickly each time or otherwise at least reminds what the used settings are each time, just before the pasting actually happens. I'm not sure why a 3rd button would be more powerful.

Apart from that, I think the "paste special" approach (selecting what to paste as you are about to paste) is quite common even outside spreadsheets (at least in the windows world?) so it might feel familiar to users.

**Comment by elstoc** [comment]
> Multi-instance though creates a problem here... Not sure how to resolve it.

That was going to be my next question

**Comment by parafin** [comment]
I don't mind any outcome either, moreover I'm doubtful my approach will be implemented (too big of a change to the core functionality). I just wanted to put it out, that current implementation is confusing, at least to me.

As for paste special - counter-argument for familiarity is that usually you select what to copy first (e.g. files in the directory) and then paste it where you want it (another directory), not the other way around. So "paste special" is quite special IMHO;) In the end I guess these 2 examples are why we now have 4 buttons - to please everybody. Which is not necessarily a good thing, since it creates confusion and complexity.

**Comment by parafin** [comment]
> > Multi-instance though creates a problem here... Not sure how to resolve it.
> 
> That was going to be my next question

Probably multi instance should be displayed in the first mode of the filter dialog (and maybe multiple history entries for the same instance, like it's done now, I think), and ignored for the second mode with all iops. Ignored meaning all instances are copied when checkbox with iop name is enabled. Otherwise it gets too compilcated:(

**Comment by elstoc** [comment]
> usually you select what to copy first (e.g. files in the directory) and then paste it where you want it (another directory), not the other way around

That's what I was going for with my suggestion. The paste parts command is the superfluous one here. Copy parts should allow you to choose a subset manually. Copy all (or whatever we're calling it now) should copy a pre-defined set of modules only (which perhaps could be user definable but by default should allow "all" or "safe"). Paste always does only one thing - paste whatever was selected by the last copy operation.

**Comment by Nilvus** [comment]
> What is the problem? That you will need to press one more button compared to current situation? You can and should use shortcuts if you're concerned with optimizing your workflow, and we already covered that it can be a single shortcut even if there are 2 buttons in the GUI.
> 
> Ability to re-filter is probably not very important, I agree, but ability to re-use the filter is IMHO quite powerful - e.g. you can copy just white balance several times (from different sources) without having to go through the dialog several times.

Reading your post and at first, I didn't understand your idea and why change it. But reading that, I totally agree and love that. Anyway, I wouldn't name that filter but indeed having one button to copy, one to paste and a filter to choose modules to copy/paste would be quite better and powerful. For me it would even be more consistent with other right modules works. For copy/paste style or even developement, we select to overwrite or stack it. That way, we choose or all modules or some modules and then copy/paste. More powerful indeed and having possibility after to save presets with which modules to copy/paste would be great and easier that way. I'm on.

**Comment by parafin** [comment]
> > usually you select what to copy first (e.g. files in the directory) and then paste it where you want it (another directory), not the other way around
> 
> That's what I was going for with my suggestion. The paste parts command is the superfluous one here.

Though if you look at how @TurboGit uses copy and paste - his first example uses paste parts.

**Comment by junkyardsparkle** [comment]
I'm not sure I've completely understood this proposal on first read, I'll read again later, but this part:

> As for paste special - counter-argument for familiarity is that usually you select what to copy first (e.g. files in the directory) and then paste it where you want it (another directory), not the other way around.

I think there *is* a general familiarity with the idea of selection as part of a copy operation, not the paste operation... but on top of that, I sometimes end up using the paste shortcut multiple times for a single copy operation, and having the additional action at the paste end doesn't really sound good for that (although tweaks to the dialog ergonomics could help with that).

**Comment by wpferguson** [comment]
I use the copy/paste history stack in 1 of 2 ways, either copy all, replace history stack, paste all, or edit an image, copy the modules I changed, append history stack, paste.  I always wondered why the selection dialog popped up when I went to paste, but I just clicked OK without thinking because it fit the paradigm that I'm used to.  @TurboGit opened my eyes to another way of using the copy/paste functionality that had never crossed my mind because I didn't think.  I normally use multiple copy/pastes to do my processing, but now I can think of a way to do one copy and multiple pastes.  Kind of a _paste if_ function.
